src/lando/api/legacy/workers/automation_worker.py

```diff
@@ -0,0 +1,295 @@
+    AutomationJob,
+)
+from lando.main.models.configuration import ConfigurationKey
+from lando.main.models.landing_job import LandingJobAction, LandingJobStatus
+from lando.main.models.repo import Repo
+from lando.main.scm.abstract_scm import AbstractSCM
+from lando.main.scm.exceptions import (
+    PatchConflict,
+    SCMInternalServerError,
+    SCMLostPushRace,
+    SCMPushTimeoutException,
+    TreeApprovalRequired,
+    TreeClosed,
+)
+from lando.utils.tasks import phab_trigger_repo_update
+
+logger = logging.getLogger(__name__)
+
+
+def map_to_pydantic_action(action_type: str, action_data: dict[str, Any]) -> Action:
+    """Convert a dict to an `Action` object.
+
+    TODO there must be a better way to do this?
+    """
+    return {
+        "add-commit": AddCommitAction,
+        "merge-onto": MergeOntoAction,
+        "tag": TagAction,
+        "add-branch": AddBranchAction,
+    }[action_type](**action_data)
+
+
+@contextmanager
+def job_processing(job: AutomationJob):
+    """Mutex-like context manager that manages job processing miscellany.
+
+    This context manager facilitates graceful worker shutdown, tracks the duration of
+    the current job, and commits changes to the DB at the very end.
+
+    Args:
+        job: the job currently being processed
+        db: active database session
+    """
+    start_time = datetime.now()
+    try:
+        yield
+    finally:
+        job.duration_seconds = (datetime.now() - start_time).seconds
+
+
+class AutomationWorker(Worker):
+    """Worker to land headless API patches."""
+
+    @property
+    def STOP_KEY(self) -> ConfigurationKey:
+        """Return the configuration key that prevents the worker from starting."""
+        return ConfigurationKey.AUTOMATION_WORKER_STOPPED
+
+    @property
+    def PAUSE_KEY(self) -> ConfigurationKey:
+        """Return the configuration key that pauses the worker."""
+        return ConfigurationKey.AUTOMATION_WORKER_PAUSED
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.last_job_finished = None
+        self.refresh_enabled_repos()
+
+    def loop(self):
+        logger.debug(
+            f"{len(self.applicable_repos)} applicable repos: {self.applicable_repos}"
+        )
+
+        # Check if any closed trees reopened since the beginning of this iteration
+        if len(self.enabled_repos) != len(self.applicable_repos):
+            self.refresh_enabled_repos()
+
+        if self.last_job_finished is False:
+            logger.info("Last job did not complete, sleeping.")
+            self.throttle(self.sleep_seconds)
+            self.refresh_enabled_repos()
+
+        with transaction.atomic():
+            job = AutomationJob.next_job(repositories=self.enabled_repos).first()
+
+        if job is None:
+            self.throttle(self.sleep_seconds)
+            return
+
+        with job_processing(job):
+            job.status = LandingJobStatus.IN_PROGRESS
```

Comment:
Is this a copy/paste error?

Reply:
What do you mean by "this"?

Reply:
this == `LandingJobStatus`, since we are dealing with `AutomationJob` objects in this worker, I expected this to say something like `AutomationJobStatus`.